### ISSUE TITLE
Implement basic delegate() function

### ISSUE THREAD

**Issue by medied** [DemocracyEarth/contracts#13]
This feature can take from PR #7

**Comment by aecc** [comment]
I'll be writing here some possible requirements for delegation:

Glossary:

- Delegation: one voter giving its voting power to another voter
- Representative: receiving part of a delegation (Representee ---> **Representative**)
- Representee: giving part of a delegation (**Representee** ---> Representative)

Initial assumptions:
- Every voter has a single unit of voting power, it can delegate it to another voter, or keep it
- Delegations are publicly visible for any voter
- No GAS optimizations are done, however it should be stated every time a decision of arguably big GAS consumption is made

Functional Requirements:
- Only registered and active voters can delegate
- Every registered and active voter can receive delegations
- For simplicity, every voter has a delegation 
- For simplicity, initially, a voter delegates to itself
- A voter delegates exactly 1 vote
- A voter can receive delegation from [1 to n] voters. n has no upper bound

Non-functional requirements:
- It's necessary to be able to fetch the status of all the delegations (voting power of every voter) just by reading the blockchain (no transaction should be made to compute the whole delegation)
- Every time a delegation is changed, a transaction should be made to the blockchain and all the delegations should be recomputed (this is probably very expensive, and unfair as it will be more costly for delegations made later than others)

**Comment by aecc** [comment]
@medied let's discuss things here that could be changed/improved from the proposed requirements, and after we agree on each point I'll make changes to the original requirements

**Comment by medied** [comment]
Looks good @aecc, glad we went over it. I went ahead and created a similar spec the same for #11

**Comment by aecc** [comment]
@medied I just added B.4 . Something that will not be implemented in this PR. I managed to make cycles work on the previous implementation of Delegation.sol . I'll create it on another issue to encourage the resolution of smaller tasks/issues

**Comment by virgile-dev** [comment]
Hey guys, @medied @aecc 

Just read this from Alessandro's post.
> Every time a delegation is changed, a transaction should be made to the blockchain and all the delegations should be recomputed (this is probably very expensive, and unfair as it will be more costly for delegations made later than others)

I was wondering if this wasn't the right time to start thinking of what happens on chain vs off chain. 
Here is a few thoughts I had written down, (Santi went over it). 

---
The client will interact with a smart contract library that will move around the vote token and settle transaction in the blockchain. We want things to move fast even if operating with a blockchain. For that we should consider what happens off-chain and on-chain.

- **Off-chain:** 
    - Delegations
        - Need to operate strictly under a logic of **Quadratic Voting** to prevent formation of monopolies that can disincentivize participation. This means that when an identity delegates tokens to another identity, the cost becomes exponentially more expensive. Alice can delegate Bob strictly 1, 2, 4, 8, 16...N tokens and no value in between.
        - When you think about them delegation are almost like payment channel in the lightning network. They require trust between parties, some degree of anonymity etc.
    - Ballot creation
    - Hashtag creation (can be flushed on-chain once attached to a ballot that received votes)
- **On-chain:**
    - Voting
        - A vote has to be connected to an identity in the system. Either integrating with Blockstack IDs or uPort IDs on Ethereum is critical in order to make the transaction politically accountable.
        - Votes are always-on and real-time, attached to an address that's connected to a proposal or idea in the system. As long as votes remain in that address they will be tallied accordingly. 
    - Revoke
        - Any user has the right to revoke votes given to other delegates or proposal having sovereign control of them.
    - Organization / districts creation

**Comment by aecc** [comment]
@virgile-dev @medied while I agree with separating responsibilities in off-chain / on-chain categories, I think we are still quite far from it. We should aim at the moment for a 100% working product on-chain. Off-chain development adds other levels of complexity, for starters there is no standard SDKs or programming languages for off-chain solutions. We need to strongly familiarize ourselves with the technology before we can start building something off-chain.
